File: frontend/app/services/ykiService.ts
import { apiRequest } from "./apiClient";
import type { ApiEnvelope } from "../state/types";

export async function startYkiSession(payload: { level_band: "A1_A2" | "B1_B2" | "C1_C2" }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: "/api/v1/yki/sessions",
    auth: true,
    body: payload,
  });
}

export async function fetchYkiSession(sessionId: string): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/yki/sessions/${sessionId}`,
    auth: true,
  });
}

export async function submitYkiAnswer(sessionId: string, payload: { answer_id: string; answer: string | boolean | number }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/answers`,
    auth: true,
    body: payload,
  });
}

export async function submitYkiWriting(sessionId: string, payload: { task_id: string; text: string }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/writing`,
    auth: true,
    body: payload,
  });
}

export async function submitYkiAudio(sessionId: string, payload: { task_id: string; audio_ref: string }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/audio`,
    auth: true,
    body: payload,
  });
}

export async function startYkiConversation(sessionId: string, payload: { task_id: string }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/speaking/conversation`,
    auth: true,
    body: payload,
  });
}

export async function submitYkiConversationTurn(
  sessionId: string,
  payload: { task_id: string; turn_id: string; audio_ref: string; transcript_text: string | null },
): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/speaking/turns`,
    auth: true,
    body: payload,
  });
}

export async function requestYkiReply(sessionId: string, payload: { task_id: string }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/speaking/reply`,
    auth: true,
    body: payload,
  });
}

export async function submitYkiExam(sessionId: string, payload: { confirm_incomplete: boolean }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/yki/sessions/${sessionId}/submit`,
    auth: true,
    body: payload,
  });
}

export async function fetchYkiCertificate(sessionId: string): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/yki/sessions/${sessionId}/certificate`,
    auth: true,
  });
}

File: backend/services/yki_service.py
from __future__ import annotations

from typing import Any

from ..services.subscription_service import require_feature
from ..voice.runtime import resolve_voice_ref
from ..yki.runtime import engine_request, get_yki_session_record, map_engine_error, store_yki_session


async def start_yki_session(*, user: dict[str, Any], payload: dict[str, Any]) -> dict[str, Any]:
    require_feature(user=user, feature="yki")
    response = await engine_request(method="POST", path="/exam/start", payload=payload)
    map_engine_error(response=response)
    runtime = response.payload
    store_yki_session(user_id=user["user_id"], runtime=runtime)
    return {"runtime": runtime}


async def get_yki_session(*, user_id: str, session_id: str) -> dict[str, Any]:
    get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(method="GET", path=f"/exam/{session_id}")
    map_engine_error(response=response)
    runtime = response.payload
    store_yki_session(user_id=user_id, runtime=runtime)
    return {"runtime": runtime}


async def submit_yki_answer(*, user_id: str, session_id: str, payload: dict[str, Any]) -> dict[str, Any]:
    get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(method="POST", path=f"/exam/{session_id}/answer", payload=payload)
    map_engine_error(response=response)
    return response.payload


async def submit_yki_writing(*, user_id: str, session_id: str, payload: dict[str, Any]) -> dict[str, Any]:
    get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(method="POST", path=f"/exam/{session_id}/writing", payload=payload)
    map_engine_error(response=response)
    return response.payload


async def submit_yki_audio(*, user_id: str, session_id: str, task_id: str, audio_ref: str) -> dict[str, Any]:
    get_yki_session_record(user_id=user_id, session_id=session_id)
    audio_file_path = resolve_voice_ref(audio_ref=audio_ref, expected_session_id=session_id, expected_task_id=task_id)
    response = await engine_request(
        method="POST",
        path=f"/exam/{session_id}/audio",
        payload={"task_id": task_id, "audio_file_path": audio_file_path},
    )
    map_engine_error(response=response)
    return response.payload


async def start_yki_conversation(*, user_id: str, session_id: str, task_id: str) -> dict[str, Any]:
    record = get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(
        method="POST",
        path="/exam/speaking/start_conversation",
        payload={"session_id": session_id, "task_id": task_id, "session_token": record["engine_session_token"]},
    )
    map_engine_error(response=response)
    return response.payload


async def submit_yki_turn(*, user_id: str, session_id: str, task_id: str, turn_id: str, audio_ref: str, transcript_text: str | None) -> dict[str, Any]:
    record = get_yki_session_record(user_id=user_id, session_id=session_id)
    audio_file_path = resolve_voice_ref(
        audio_ref=audio_ref,
        expected_session_id=session_id,
        expected_task_id=task_id,
        expected_turn_id=turn_id,
    )
    response = await engine_request(
        method="POST",
        path="/exam/speaking/submit_turn",
        payload={
            "session_id": session_id,
            "task_id": task_id,
            "turn_id": turn_id,
            "audio_file_path": audio_file_path,
            "session_token": record["engine_session_token"],
            "transcript_text": transcript_text,
        },
    )
    map_engine_error(response=response)
    return response.payload


async def generate_yki_reply(*, user_id: str, session_id: str, task_id: str) -> dict[str, Any]:
    record = get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(
        method="POST",
        path="/exam/speaking/generate_reply",
        payload={"session_id": session_id, "task_id": task_id, "session_token": record["engine_session_token"]},
    )
    map_engine_error(response=response)
    return response.payload


async def submit_yki_exam(*, user_id: str, session_id: str, confirm_incomplete: bool) -> dict[str, Any]:
    record = get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(
        method="POST",
        path=f"/exam/{session_id}/submit",
        payload={"confirm_incomplete": confirm_incomplete, "session_token": record["engine_session_token"]},
    )
    map_engine_error(response=response)
    return response.payload


async def get_yki_certificate(*, user_id: str, session_id: str) -> dict[str, Any]:
    get_yki_session_record(user_id=user_id, session_id=session_id)
    response = await engine_request(method="GET", path=f"/exam/{session_id}/certificate")
    map_engine_error(response=response)
    return response.payload
